tsfc/loopy.py

```diff
@@ -0,0 +1,338 @@
+        statements.extend(statement(child, ctx))
+    return statements
+
+
+@statement.register(imp.For)
+def statement_for(tree, ctx):
+    extent = tree.index.extent
+    assert extent
+    idx = ctx.name_gen(ctx.index_names[tree.index])
+    ctx.active_indices[tree.index] = p.Variable(idx)
+    ctx.index_extent[idx] = extent
+
+    statements = statement(tree.children[0], ctx)
+
+    ctx.active_indices.pop(tree.index)
+    return statements
+
+
+@statement.register(imp.Initialise)
+def statement_initialise(leaf, ctx):
+    return [lp.Assignment(expression(leaf.indexsum, ctx), 0.0, within_inames=ctx.active_inames())]
+
+
+@statement.register(imp.Accumulate)
+def statement_accumulate(leaf, ctx):
+    lhs = expression(leaf.indexsum, ctx)
+    rhs = lhs + expression(leaf.indexsum.children[0], ctx)
+    return [lp.Assignment(lhs, rhs, within_inames=ctx.active_inames())]
+
+
+@statement.register(imp.Return)
+def statement_return(leaf, ctx):
+    lhs = expression(leaf.variable, ctx)
+    rhs = lhs + expression(leaf.expression, ctx)
+    return [lp.Assignment(lhs, rhs, within_inames=ctx.active_inames())]
+
+
+@statement.register(imp.ReturnAccumulate)
+def statement_returnaccumulate(leaf, ctx):
+    lhs = expression(leaf.variable, ctx)
+    rhs = lhs + expression(leaf.indexsum.children[0], ctx)
+    return [lp.Assignment(lhs, rhs, within_inames=ctx.active_inames())]
+
+
+@statement.register(imp.Evaluate)
+def statement_evaluate(leaf, ctx):
+    expr = leaf.expression
+    if isinstance(expr, gem.ListTensor):
+        ops = []
+        var = ctx.pymbolic_variable(expr)
+        index = ()
+        if isinstance(var, p.Subscript):
+            var, index = var.aggregate, var.index_tuple
+        for multiindex, value in numpy.ndenumerate(expr.array):
+            ops.append(lp.Assignment(p.Subscript(var, index + multiindex), expression(value, ctx), within_inames=ctx.active_inames()))
+        return ops
+    elif isinstance(expr, gem.Constant):
+        return []
+    else:
+        return [lp.Assignment(ctx.pymbolic_variable(expr), expression(expr, ctx, top=True), within_inames=ctx.active_inames())]
+
+
+def expression(expr, ctx, top=False):
+    """Translates GEM expression into a pymbolic expression
+
+    :arg expr: GEM expression
+    :arg ctx: miscellaneous code generation data
+    :arg top: do not generate temporary reference for the root node
+    :returns: pymbolic expression
+    """
+    if not top and expr in ctx.gem_to_pymbolic:
+        return ctx.pymbolic_variable(expr)
+    else:
+        return _expression(expr, ctx)
+
+
+@singledispatch
+def _expression(expr, parameters):
+    raise AssertionError("cannot generate expression from %s" % type(expr))
+
+
+@_expression.register(gem.Failure)
+def _expression_failure(expr, parameters):
+    raise expr.exception
+
+
+@_expression.register(gem.Product)
+def _expression_product(expr, ctx):
+    return p.Product(tuple([expression(c, ctx) for c in expr.children]))
```

Comment:
No need to cons a list, why not: `tuple(expression(c, ctx) for c in expr.children))`

Reply:
Oh yes!